src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/ItemManager.cs

```diff
@@ -0,0 +1,477 @@
+        {
+            var highlightedList = new List<AsyncCompletionData.CompletionItemWithHighlight>();
+            foreach (var item in filterResults)
+            {
+                var highlightedSpans = highlightMatchingPortions
+                                        ? _completionHelper.GetHighlightedSpans(item.VSCompletionItem.DisplayText, filterText, CultureInfo.CurrentCulture)
+                                        : ImmutableArray<TextSpan>.Empty;
+                highlightedList.Add(new AsyncCompletionData.CompletionItemWithHighlight(item.VSCompletionItem, highlightedSpans.Select(s => s.ToSpan()).ToImmutableArray()));
+            }
+
+            return highlightedList;
+        }
+
+        private static ImmutableArray<AsyncCompletionData.CompletionFilterWithState> GetUpdatedFilters(
+            List<ExtendedFilterResult> filteredList,
+            ImmutableArray<AsyncCompletionData.CompletionFilterWithState> filters)
+        {
+            // See which filters might be enabled based on the typed code
+            var textFilteredFilters = filteredList.SelectMany(n => n.VSCompletionItem.Filters).Distinct();
+
+            // When no items are available for a given filter, it becomes unavailable
+            return ImmutableArray.CreateRange(filters.Select(n => n.WithAvailability(textFilteredFilters.Contains(n.Filter))));
+        }
+
+        private void MakeMostRecentItem(string item)
+        {
+            var updated = false;
+
+            while (!updated)
+            {
+                var oldItems = _recentItems;
+                var newItems = oldItems.Remove(item);
+
+                if (newItems.Length == MaxMRUSize)
+                {
+                    // Remove the least recent item.
+                    newItems = newItems.RemoveAt(0);
+                }
+
+                newItems = newItems.Add(item);
+                updated = ImmutableInterlocked.InterlockedCompareExchange(ref _recentItems, newItems, oldItems) == oldItems;
+            }
+        }
+
+        private static bool ShouldBeFilteredOutOfCompletionList(VSCompletionItem item, ImmutableArray<AsyncCompletionData.CompletionFilter> activeFilters)
+        {
+            foreach (var itemFilter in item.Filters)
+            {
+                if (activeFilters.Contains(itemFilter))
+                {
+                    return false;
+                }
+            }
+
+            return true;
+        }
+
+        private void ItemCommitted(object sender, AsyncCompletionData.CompletionItemEventArgs e)
+        {
+            MakeMostRecentItem(e.Item.DisplayText);
+        }
+
+        private void SessionDismissed(object sender, EventArgs e)
+        {
+            if (sender is IAsyncCompletionSession session)
+            {
+                UnsubscribeEvents(session);
+            }
+        }
+
+        private void SubscribeEvents(IAsyncCompletionSession session)
+        {
+            UnsubscribeEvents(session);
```

Comment:
Can we have a session here that we would need to unsubscribe from?  how does that happen?  that indicates to me that someone is calling SbscribeEvents before telling us the previous session was dismissed.  that seems *bad*.

Reply:
Replaced this with a singletone. Thanks!

---
In reply to: [232510747](https://github.com/dotnet/roslyn/pull/29016#discussion_r232510747) [](ancestors = 232510747)